### PR TITLE
put test objects in implements and test modules

### LINKED ISSUE #35: Move test objects back to the `@implements` macro
Moving the data to the tests was a mistake. We lose all high level approaches to testing.

The context for this is the BaseInterfaces.jl tests are getting ridiculous - its hard to track which interfaces are actually tested.

It should be literally just:

```julia
Interfaces.test(BaseInterfaces)
```

Then we look through the `implements` method table and test everything in there defined by BaseInterfaces.jl

But that requires having test data attached to implementations.

(Note we should still be able to pass objects to the tests, we just shouldn't have to)

### PR DESCRIPTION
**Pull request by rafaqz**
This PR moves test objects back to the `@implements` macro, and implements the `test` method for Module:

```julia
@test Interfaces.test(SomeModule)
```

Which cuts a huge amount of code from BaseInterfaces.jl and ensures everything implemented is tested.

Closes #35

### CODE REVIEW

File: BaseInterfaces/src/implementations.jl [hunk 1]
```diff
@@ -1,43 +1,52 @@
...
+@implements IterationInterface{(:reverse,:indexing)} UnitRange [1:5, -2:2]
+@implements IterationInterface{(:reverse,:indexing)} StepRange [1:2:10, 20:-10:-20]
+@implements IterationInterface{(:reverse,:indexing)} Array [[1, 2, 3, 4], [:a :b; :c :d]]
+@implements IterationInterface{(:reverse,:indexing)} Tuple [(1, 2, 3, 4)]
+@static if VERSION >= v"1.9.0"
+    @implements IterationInterface{(:reverse,:indexing)} NamedTuple [(a=1, b=2, c=3, d=4)]
+else
+    @implements IterationInterface{:indexing} NamedTuple [(a=1, b=2, c=3, d=4)] # No reverse on 1.6
+end
+# @implements IterationInterface{(:reverse,:indexing)} String
```
Comment (gdalle) [inline BaseInterfaces/src/implementations.jl:37]: What's up with the commented ones?

File: BaseInterfaces/test/runtests.jl [hunk 2]
```diff
@@ -2,55 +2,16 @@ using BaseInterfaces
...
-    @test Interfaces.test(ArrayInterface, UnitRange, [2:10])
-    @test Interfaces.test(ArrayInterface, StepRange, [2:1:10])
-    @test Interfaces.test(ArrayInterface, Base.OneTo, [Base.OneTo(10)])
-    @test Interfaces.test(ArrayInterface, Base.Slice, [Base.Slice(100:150)])
-    @test Interfaces.test(ArrayInterface, Base.IdentityUnitRange, [Base.IdentityUnitRange(100:150)])
-    @test Interfaces.test(ArrayInterface, Base.CodeUnits, [codeunits("abcde")])
-    # No `getindex` defined for LogicalIndex
-    @test_broken Interfaces.test(ArrayInterface, Base.LogicalIndex, [to_indices([1, 2, 3], ([false, true, true],))[1]])
+# Test all interfaaces in BaseInterfaces
+@test Interfaces.test(BaseInterfaces)
```
Comment (gdalle) [inline BaseInterfaces/test/runtests.jl:9]: Does this test all interfaces _defined_ in the module, or _implemented_ by the module?

File: BaseInterfaces/test/runtests.jl [hunk 3]
```diff
@@ -2,55 +2,16 @@ using BaseInterfaces
 using Interfaces
 using Test
 
-@implements SetInterface{(:empty,:emptymutable,:hasfastin,:intersect,:union,:sizehint!)} Test.GenericSet
-@implements DictInterface Test.GenericDict
+@implements SetInterface{(:empty,:emptymutable,:hasfastin,:intersect,:union,:sizehint!)} Test.GenericSet [Test.GenericSet(Set((1, 2)))]
```
Comment (gdalle) [inline BaseInterfaces/test/runtests.jl:5]: Why do you need those in addition to `Interfaces.test(BaseInterfaces)`?

File: src/implements.jl [hunk 4]
```diff
@@ -40,14 +40,14 @@ using BaseInterfaces
...
 ```
 """
-macro implements(interface, objtype)
-    _implements_inner(interface, objtype)
+macro implements(interface, objtype, test_objects)
+    _implements_inner(interface, objtype, test_objects)
 end
-macro implements(dev::Symbol, interface, objtype)
+macro implements(dev::Symbol, interface, objtype, test_objects)
     dev == :dev || error("3 arg version of `@implements must start with `dev`, and should only be used in testing")
```
Comment (gdalle) [inline src/implements.jl:47]: I don't think `dev` does anything anymore, does it?

File: src/implements.jl [hunk 5]
```diff
@@ -40,14 +40,14 @@ using BaseInterfaces
...
-    _implements_inner(interface, objtype; show=true)
+    _implements_inner(interface, objtype, test_objects; show=true)
 end
-function _implements_inner(interface, objtype; show=false)
+function _implements_inner(interface, objtype, test_objects; show=false)
```
Comment (gdalle) [inline src/implements.jl:50]: So the test does _not_ happen when we run `@implements`?

File: src/test.jl [hunk 6]
```diff
@@ -41,23 +41,54 @@ function check_coherent_types(O::Type, tow::TestObjectWrapper)
...
-    test(::Type{<:Interface}, obj)
+    test(m::Module)
+    test(::Type{<:Interface}, m::Module)
+    test(::Type{<:Interface}, obj::Type)
 
 Test if an interface is implemented correctly for an object,
 returning `true` or `false`.
 
 If no interface type is passed, Interfaces.jl will find all the
 interfaces available and test them.
```
Comment (gdalle) [inline src/test.jl:52]: Defined in the module or implemented by types from the module?

File: src/test.jl [hunk 7]
```diff
@@ -41,23 +41,54 @@ function check_coherent_types(O::Type, tow::TestObjectWrapper)
...
+    methodlist = methods(implements, Tuple{T,<:Any})
+    _test_module(mod, methodlist; kw...)
+end
+function test(mod::Module; kw...) 
+    methodlist = methods(implements, Tuple{<:Any,<:Any})
+    _test_module(mod, methodlist; kw...)
+end
+
+
+function _test_module(mod, methodlist; kw...)
```
Comment (gdalle) [inline src/test.jl:64]: I have zero clue what this function does. Send help

File: src/test.jl [hunk 8]
```diff
@@ -118,7 +149,9 @@ function _test(T, name::Symbol, condition, obj, i=nothing)
     obj_copy = deepcopy(obj)
     res = try
         f = condition isa Pair ? condition[2] : condition
+        # GC.enable(false)
```
Comment (gdalle) [inline src/test.jl:152]: What's up with that?

File: BaseInterfaces/test/runtests.jl [hunk 9]
```diff
@@ -2,55 +2,16 @@ using BaseInterfaces
...
-
-@testset "SetInterface" begin
-    @test Interfaces.test(SetInterface, Set, [Set((1, 2))])
-    @test Interfaces.test(SetInterface, BitSet, [BitSet((1, 2))])
-    @test Interfaces.test(SetInterface, Base.KeySet, [Base.KeySet(Dict(:a=>1, :b=>2))])
-    @test Interfaces.test(SetInterface, Test.GenericSet, [Test.GenericSet(Set((1, 2)))])
-    s = Base.IdSet(); push!(s, "a"); push!(s, "b")
-    @test Interfaces.test(SetInterface, Base.IdSet, [s]) 
-end
+@test_broken Interfaces.test(ArrayInterface, Base.LogicalIndex, [to_indices([1, 2, 3], ([false, true, true],))[1]])
```
Comment (gdalle) [inline BaseInterfaces/test/runtests.jl:17]: Is this fixable in Base Julia?

File: BaseInterfaces/src/implementations.jl (hunk 1 above)
Comment (rafaqz) [reply BaseInterfaces/src/implementations.jl:37]: They weren't actually tested before anyway :sweat_smile: 

Good reason for the PR in itself. Guess I should make some test data for them...

File: src/implements.jl (hunk 4 above)
Comment (rafaqz) [reply src/implements.jl:47]: I can't even remember, but we should delete it

File: BaseInterfaces/test/runtests.jl (hunk 3 above)
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:5]: Oh yeah they will be in `Main`.

File: src/implements.jl (hunk 5 above)
Comment (rafaqz) [reply src/implements.jl:50]: No, it just puts the expression to generate the objects into `test_objects(T, O) = $test_objects` for us to get later

File: src/test.jl (hunk 6 above)
Comment (rafaqz) [reply src/test.jl:52]: Yeah I'll clarify the doc

File: src/test.jl (hunk 7 above)
Comment (rafaqz) [reply src/test.jl:64]: Ok, I'll add a comment! But yes this is where we dig into some internals and use the method table as the global state of all available implementations.

Its looping over the methods in the methodlist from `methods(implementes `Tuple{T, <:Any})` and checking which of them the `implements` method is in `mod`, then running `test` on the type signature, that we have to reconstruct a little.

File: src/test.jl (hunk 8 above)
Comment (rafaqz) [reply src/test.jl:152]: Yeah, remnant of tryng to fix `WeakRefDict` - GC collects the refs before we can test with them.

File: BaseInterfaces/test/runtests.jl (hunk 9 above)
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:17]: Yes by putting `collect(LogicalIndex(I))` [here](https://github.com/JuliaLang/julia/blob/2a842146ac6ee0a7f505d56b4c1f77274d7f0628/base/indices.jl#L294C39-L294C39)

But of course that has a serious performance hit.

See: https://github.com/JuliaLang/julia/issues/51071

File: BaseInterfaces/test/runtests.jl (hunk 2 above)
Comment (rafaqz) [reply BaseInterfaces/test/runtests.jl:9]: Should be everything implemented in the module if its working... The specific test is if the method of `implements` is defined in the module.

**Pull request merged**